Step3 - 인증을 통한 기능 구현 리뷰 요청드립니다.

programmer-sjk opened the pull request:
안녕하세요 동휘님.

이번 미션은 3가지 요구사항을 만족해야 하다 보니 양이 많았다는 느낌이 들었습니다.
참, 이번에 생성자 접근제어 수정과 final 키워드를 추가해서 File changes가 조금 많습니다.

그럼 리뷰 부탁드릴게요 🙏

programmer-sjk committed:
docs(guide): README.md 요구사항 정리

programmer-sjk committed:
feat(service): 토큰이 유효하지 않을 경우 AuthorizationException 발생

programmer-sjk committed:
test(acceptance): 토큰 발급 기능(로그인) 인수 테스트 추가

programmer-sjk committed:
refactor(test): MemberRestAssured로 통신하는 함수분리

programmer-sjk committed:
test(acceptance): 내 정보 조회/수정/삭제 RestAssured 함수 추가

programmer-sjk committed:
test(acceptance): 나의 정보를 관리하는 인수테스트 추가

programmer-sjk committed:
feat(controller): 내 정보 관리 인수테스트가 성공하도록 AuthenticationPrincipal 어노테이션 추가

programmer-sjk committed:
feat(acceptance): Auth 인수테스트에 유효하지 않은 토큰에 대한 테스트 성공하도록 수정

programmer-sjk committed:
refactor(service): AuthService 멤버 없을 때 예외처리 리팩토링

programmer-sjk committed:
refactor: 코드 전반적인 컨벤션 리팩토링

리팩토링
- final 키워드 추가
- public 생성자를 private 수정

programmer-sjk committed:
test(acceptance): 즐겨찾기 RestAssured 함수 추가

programmer-sjk committed:
test(acceptance): 즐겨찾기 인수테스트 작성

programmer-sjk committed:
test(domain): Favorite 도메인 테스트 추가

programmer-sjk committed:
feat(domain): Favorite 도메인 추가

programmer-sjk committed:
test(fixture): Member 생성관련 텍스트 픽스쳐 추가

programmer-sjk committed:
feat(repository): Favorite repository 추가

programmer-sjk committed:
test(service): Favorite service 테스트 추가

programmer-sjk committed:
refactor(repository): Favorite 레포지토리에 findAllByMemberId 선언 추가

programmer-sjk committed:
refactor: FavoriteResponse에 from 정적 팩토리 메소드 추가

programmer-sjk committed:
feat(service): FavoriteService 코드 구현

programmer-sjk committed:
feat(controller): Favorite Controller 추가

programmer-sjk committed:
refactor: 접근제어자 수정, 안쓰는 import 문 정리

programmer-sjk committed:
refactor: 에러 문자열 대신 Enum 활용

programmer-sjk committed:
refactor: MemberService 리팩토링 및 그에 따라 깨지는 테스트 수정

dhmin5693 commented on src/main/java/nextstep/subway/favorite/ui/FavoriteController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @DeleteMapping("/{id}")
+    public ResponseEntity<Void> deleteFavorite(@PathVariable Long id) {
```
미션 요강을 보시면 삭제도 authorization 헤더를 요구하고 있습니다.

programmer-sjk replied:
아앗...!!! 제가 놓쳤군요 감사합니다. ㅠㅠ

dhmin5693 commented on src/main/java/nextstep/subway/favorite/dto/FavoriteResponse.java:
```diff
@@ -0,0 +1,38 @@
+public class FavoriteResponse {
+    Long id;
+    StationResponse source;
+    StationResponse target;
```
```suggestion
    private Long id;
    private StationResponse source;
    private StationResponse target;
```

접근 제한자가 빠진 것 같아요!

dhmin5693 commented on src/test/java/nextstep/subway/favorite/acceptance/FavoriteAcceptanceTest.java:
```diff
@@ -0,0 +1,101 @@
+                .getAccessToken();
+    }
+
+    @DisplayName("즐겨찾기를 생성할 수 있다")
+    @Test
+    void createFavorite() {
+        // given
+        FavoriteRequest request = new FavoriteRequest(양재역.getId(), 연신내역.getId());
+
+        // when
+        ExtractableResponse<Response> response = 즐겨찾기_생성_요청(token, request);
+
+        // then
+        즐겨찾기_등록됨(response);
+    }
```
https://tecoble.techcourse.co.kr/post/2020-07-31-dynamic-test/

테스트 메소드는 request, API 실행, 결과 검증 순서대로인 것처럼 보이지만 로직 수행 전 로그인을 시도하고 있습니다.
TestFactory를 활용하면 일련의 테스트를 가독성 있게 작성하는 것이 가능합니다.

지금 굳이 하실 필요는 없지만 적용해볼만한 곳이 있다면 한번 해보시는걸 추천드립니다.

dhmin5693 reviewed (changes requested):
요구사항이 많아서 그런지 코드 수정량도 되게 많네요 🥲
전반적으로 잘 작성해주셔서 스펙만 일치하면 바로 머지 눌러도 될 것 같아요.

남긴 피드백 확인해보시고 다시 요청주세요 😁

programmer-sjk committed:
refactor: 코드 리뷰 반영

코드 리뷰 반영
- 즐겨찾기 삭제 시, 자신의 즐겨찾기만 삭제하도록 수정
- 타인의 즐겨찾기 삭제 할수 없는 테스트 추가
- 누락한 접근제한자 추가
- dynamic 테스트 적용

programmer-sjk commented:
안녕하세요 동휘님 제가 놓친 부분을 꼼꼼히 봐주셔서 감사합니다 ㅠㅠ.

**타인의 즐겨찾기를 삭제할 수 없는 테스트 추가**하였고 `dynamic-test`는 Member 인수테스트의 `manageMyInfo` 에 적용해 봤습니다.
dynamic test는 처음 써보는데 혹시 피드백 주시고 싶은 부분 있으면 편하게 남겨주세요!

dhmin5693 commented on src/test/java/nextstep/subway/member/MemberAcceptanceTest.java:
```diff
@@ -46,54 +52,43 @@ void manageMember() {
-                .then().log().all()
-                .extract();
+    @TestFactory
+    Stream<DynamicTest> manageMyInfo() {
```
👍 
이런 방식으로 사용하시면 됩니다.
순서가 중요한 연속적인 테스트를 작성할 때 꽤나 유용해요.

dhmin5693 reviewed (approved):
피드백 빠르게 반영해주셨군요 😄
4단계 미션 진행해주세요~
여기도 요구사항이 많이 복잡해서 어렵지만 조금만 더 힘내시길 바랍니다 💪💪
